File: 백준/Silver/2579. 계단 오르기/계단 오르기.cc
#include <iostream>
#include <string>
#include <vector>

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);

using namespace std;

int main(){
    FASTIO;
    
    int n;
    cin >> n;

    vector<int> point(n + 1, 0);
    for(int i = 1; i <= n; i++){
        cin >> point[i];
    }

    vector<vector<int>> stair(n + 1, vector<int>(2, 0));
    stair[1][0] = point[1];
    stair[1][1] = point[1];

    for(int i = 2; i <= n; i++){
        stair[i][0] = max(stair[i - 2][0], stair[i - 2][1]) + point[i];
        stair[i][1] = stair[i - 1][0] + point[i];
    }

    cout << max(stair[n][0], stair[n][1]) << endl;
    
    return 0;
}

File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
import sys

n = int(sys.stdin.readline())
stair = [0]
for _ in range(n):
    stair.append(int(sys.stdin.readline()))

point = [[0, 0] for _ in range(n + 1)]
# point[i][0] = 구간 [1, i]에서 stair[i]를 이전 계단과 연속하지 않고 밟을 때의 점수 최댓값
# point[i][1] = 구간 [1, i]에서 stair[i]를 이전 계단과 연속해서 밟을 때의 점수 최댓값
point[1] = [stair[1], stair[1]]

for i in range(2, n + 1):
    point[i][0] = max(point[i - 2]) + stair[i]
    point[i][1] = point[i - 1][0] + stair[i]

print(max(point[n]))